src/main/java/it/gov/pagopa/reward/notification/service/iban/outcome/recovery/CompletedKoDiscardedRewardNotificationServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+                        .flatMap(this::createRemedialNotification)
+                        .collectList(),
+                recovered -> "Recovered %d COMPLETED_KO rewardNotification on userId %s and initiativeId %s".formatted(
+                        recovered.size(),
+                        rewardIban.getUserId(),
+                        rewardIban.getInitiativeId()));
+    }
+
+    private Mono<RewardsNotification> createRemedialNotification(RewardsNotification discarded) {
+        log.info("[REWARD_NOTIFICATION_IBAN_OUTCOME] Found discarded COMPLETED_KO rewardNotification having id {} on userId {} and initiativeId {}",
+                discarded.getId(), discarded.getUserId(), discarded.getInitiativeId());
+
+        return Mono.just(discarded)
+                .flatMap(this::buildRemedialNotification)
+                .flatMap(remedialNotification -> updateDiscardedAndStoreRemedial(discarded, remedialNotification))
+
+                .onErrorResume(e -> {
+                    log.error("[REWARD_NOTIFICATION_IBAN_OUTCOME] Something went wrong while recovering COMPLETED_KO rewardNotification having id {} related to userId {} and initiativeId {}",
+                            discarded.getId(), discarded.getUserId(), discarded.getInitiativeId(), e);
+                    return Mono.empty();
+                });
+    }
+
+    private Mono<RewardsNotification> buildRemedialNotification(RewardsNotification discarded) {
+
+        RewardsNotification remedial = discarded.toBuilder()
+                .status(RewardNotificationStatus.TO_SEND)
+                .exportId(null)
+                .exportDate(null)
+                .iban(null)
+                .rejectionReason(null)
+                .resultCode(null)
+                .feedbackDate(null)
+                .feedbackHistory(Collections.emptyList())
+                .cro(null)
+                .executionDate(null)
+                .remedialId(null)
+                .build();
+
+        setNewIdAndOrdinaryId(remedial, discarded);
+
+
+        return setRemedialNotificationDate(discarded.getInitiativeId(), remedial);
+    }
+
+    private void setNewIdAndOrdinaryId(RewardsNotification remedial, RewardsNotification discarded) {
+        String id = discarded.getId();
+        String externalId = discarded.getExternalId();
+
+        // if recovering a remedial it will calculate the next progressive
+        if (discarded.getOrdinaryId() != null) {
+
+            remedial.setId(buildRecoveryId(id, discarded.getOrdinaryId()));
+            remedial.setExternalId(buildRecoveryId(externalId, discarded.getOrdinaryExternalId()));
+
+            remedial.setOrdinaryId(discarded.getOrdinaryId());
+            remedial.setOrdinaryExternalId(discarded.getOrdinaryExternalId());
+        } else {
+            remedial.setId(id.concat("%s1".formatted(RECOVERY_ID_SUFFIX)));
+            remedial.setExternalId(externalId.concat("%s1".formatted(RECOVERY_ID_SUFFIX)));
```

Comment:
you could centralize also these 2 lines with a unique method (buildFirstRecoveryId)
where instead of use concat together with a formatted, you can use just a formatted